### PR TITLE
[Bounty] ConCare PyHealth 2.0 ver

### PR DESCRIPTION
**Pull request by joshuasteier**
<html>
<body>
<html><head></head><body><h1>Pull Request: ConCare Model Update (1.0 → 2.0)</h1>
<h2>Contributor Information</h2>
<ul>
<li><strong>Name:</strong> Joshua Steier</li>
</ul>
<h2>Contribution Type</h2>
<ul>
<li>[x] Model Update</li>
<li>[ ] New Model</li>
<li>[ ] Dataset</li>
<li>[ ] Task</li>
<li>[ ] Bug Fix</li>
<li>[ ] Documentation</li>
<li>[ ] Other</li>
</ul>
<h2>High-Level Description</h2>
<p>This PR updates the <strong>ConCare model</strong> from PyHealth 1.0 API to PyHealth 2.0 API.</p>
<h3>Paper Reference</h3>
<ul>
<li><strong>Title:</strong> Concare: Personalized clinical feature embedding via capturing the healthcare context</li>
<li><strong>Authors:</strong> Liantao Ma et al.</li>
<li><strong>Venue:</strong> AAAI 2020</li>
<li><strong>Link:</strong> https://ojs.aaai.org/index.php/AAAI/article/view/5428</li>
</ul>
<h3>Changes Made</h3>
<ol>
<li>
<p><strong>API Migration (1.0 → 2.0):</strong></p>
<ul>
<li>Replaced <code>SampleEHRDataset</code> with <code>SampleDataset</code></li>
<li>Integrated <code>EmbeddingModel</code> for unified embedding handling</li>
<li>Removed explicit <code>feature_keys</code>, <code>label_key</code>, <code>mode</code>, <code>use_embedding</code> parameters</li>
<li>Simplified constructor to derive feature information from dataset schemas</li>
</ul>
</li>
<li>
<p><strong>Code Improvements:</strong></p>
<ul>
<li>Added comprehensive Google-style docstrings for all classes and methods</li>
<li>Added proper type hints throughout</li>
<li>Fixed bug in <code>SingleAttention</code> (removed undefined <code>self.Wd</code> initialization)</li>
<li>Improved code documentation with input/output descriptions</li>
<li>Added file header with paper reference and description</li>
</ul>
</li>
<li>
<p><strong>Testing:</strong></p>
<ul>
<li>Added comprehensive unit tests covering:
<ul>
<li>Model initialization (with/without static features)</li>
<li>Forward pass validation</li>
<li>Backward pass (gradient flow)</li>
<li>Embedding extraction</li>
<li>Custom hyperparameters</li>
<li>Multiclass classification</li>
<li>Single feature input</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong>Examples:</strong></p>
<ul>
<li>Added example notebook for MIMIC-IV in-hospital mortality prediction</li>
<li>Added standalone Python script example</li>
</ul>
</li>
</ol>
<h2>Files to Review</h2>

File | Description
-- | --
pyhealth/models/concare.py | Main model implementation (updated)
tests/models/test_concare.py | Unit tests for ConCare model
examples/concare_mimic4_example.ipynb | Example notebook for MIMIC-IV

<h2>How to Test</h2>
<h3>1. Run Unit Tests</h3>
<pre><code class="language-bash">python -m pytest tests/models/test_concare.py -v
</code></pre>
<h3>2. Run Quick Example (main block)</h3>
<pre><code class="language-bash">python pyhealth/models/concare.py
</code></pre>
<h3>3. Expected Output</h3>
<pre><code class="language-python">{
    'loss': tensor(..., grad_fn=&lt;AddBackward0&gt;),
    'y_prob': tensor([[...], [...]], grad_fn=&lt;SigmoidBackward0&gt;),
    'y_true': tensor([[...], [...]]),
    'logit': tensor([[...], [...]], grad_fn=&lt;AddmmBackward0&gt;)
}
</code></pre>
<h2>Checklist</h2>
<ul>
<li>[x] Code follows PEP8 style (88 character line length)</li>
<li>[x] Code follows Google-style docstrings</li>
<li>[x] All functions have type hints</li>
<li>[x] All functions have input/output documentation</li>
<li>[x] File header includes author, paper title, link, and description</li>
<li>[x] Unit tests pass</li>
<li>[x] Example code runs successfully</li>
<li>[x] Backward pass (loss.backward()) works</li>
<li>[x] Code is rebased with main branch</li>
</ul>
<h2>Additional Notes</h2>
<p>The ConCare model includes:</p>
<ul>
<li><strong>Channel-wise GRUs:</strong> Separate GRU for each input feature dimension</li>
<li><strong>Time-aware attention:</strong> Captures temporal decay in healthcare context</li>
<li><strong>Multi-head self-attention:</strong> Captures feature interactions</li>
<li><strong>DeCov loss:</strong> Regularization to reduce feature redundancy</li>
<li><strong>Static feature support:</strong> Optional demographic/static features</li>
</ul>
<p>This update maintains full backward compatibility with the original ConCare functionality while adopting the cleaner 2.0 API patterns.</p></body></html>
</body>
</html>

### CODE REVIEW

File: examples/concare_mimic4_example.ipynb
Comment (jhnwu3) [inline examples/concare_mimic4_example.ipynb:1]: Will review more of the codebase later (as I find time tomorrow sorry)

For some reason this notebook doesn't show up in GitHub?  "The Notebook Does Not Appear to Be Valid JSON"

Comment (joshuasteier) [reply examples/concare_mimic4_example.ipynb:1]: Thank you, no pressure.
I believe I've fixed this error now in the latest commit.
Much appreciated for your guidance.

File: pyhealth/models/concare.py [hunk 1]
```diff
@@ -547,415 +724,289 @@ def concare_encoder(self, input, static=None, mask=None):
...
+                        static_val.shape[-1] if static_val.dim() > 0 else 1
+                    )
+                elif isinstance(static_val, (list, tuple)):
+                    self.static_dim = len(static_val)
+                else:
+                    self.static_dim = 1
+
+        # Get dynamic feature keys (excluding static key)
+        self.dynamic_feature_keys = [
+            k for k in self.dataset.input_processors.keys()
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/concare.py]: Fixed - updated to use self.feature_keys

File: pyhealth/models/concare.py [hunk 2]
```diff
@@ -547,415 +724,289 @@ def concare_encoder(self, input, static=None, mask=None):
...
-                    kwargs[self.static_key], dtype=torch.float, device=self.device
+                    static_data, dtype=torch.float, device=self.device
                 )
-                x, decov = self.concare[feature_key](x, static=static, mask=mask)
-            else:
-                x, decov = self.concare[feature_key](x, mask=mask)
+
+        for feature_key in self.dynamic_feature_keys:
+            x = embedded[feature_key]
+            mask = (x.sum(dim=-1) != 0).int()
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/concare.py]: Fixed - now using output_mask=True

File: pyhealth/models/concare.py [hunk 3]
```diff
@@ -547,415 +724,289 @@ def concare_encoder(self, input, static=None, mask=None):
...
-        self.label_tokenizer = self.get_label_tokenizer()
-        # the key of self.embeddings only contains the code based inputs
-        self.embeddings = nn.ModuleDict()
-        # the key of self.linear_layers only contains the float/int based inputs
-        self.linear_layers = nn.ModuleDict()
+        assert len(self.label_keys) == 1, (
+            "Only one label key is supported for ConCare"
+        )
+        self.label_key = self.label_keys[0]
+        self.mode = self.dataset.output_schema[self.label_key]
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/concare.py]: Fixed - removed redundant self.mode

File: pyhealth/models/concare.py [hunk 4]
```diff
@@ -166,102 +249,189 @@ def __init__(self, h, d_model, dropout=0):
...
-    def __init__(self, size, dropout):
+    def __init__(self, size: int, dropout: float):
         super(SublayerConnection, self).__init__()
         self.norm = LayerNorm(size)
         self.dropout = nn.Dropout(dropout)
 
-    def forward(self, x, sublayer):
-        "Apply residual connection to any sublayer with the same size."
+    def forward(
+        self, x: torch.Tensor, sublayer: callable
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/concare.py]: Fixed - added Callable to imports

File: pyhealth/models/concare.py [hunk 5]
```diff
@@ -1,16 +1,60 @@
...
+    - Multi-head self-attention for capturing feature interactions
+    - DeCov regularization loss to reduce feature redundancy
+    - Final attention mechanism for patient-level representation
+
+    This implementation has been updated from PyHealth 1.0 to PyHealth 2.0 API,
+    using the new SampleDataset and EmbeddingModel interfaces.
+"""
+
 import math
 from typing import Dict, List, Optional, Tuple
```
[the remark this answers is not in the document]
Comment (joshuasteier) [reply pyhealth/models/concare.py]: Fixed - added Callable to imports

File: examples/concare_mimic4_example.ipynb
Comment (joshuasteier) [reply examples/concare_mimic4_example.ipynb:1]: @jhnwu3, been updated to fix any issues. Thank you!

**Review by jhnwu3 — APPROVED** [review]
Clean adaptation, no complaints for me.
